docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+* user needs to specify the exact path, thus requires knowledge of the filesystem
+* hard to dynamically provision a hostpath
+* hostpath volumes are not cleaned up
+
+The `emptydir` abstraction solves these problems:
+* it is allocated in `/var/lib/kubelet/resource/uid/`, a location that is predictable but unknow till `resource` creation time
+* The node takes care of allocation, a dynamic provisioner just needs to create the PV
+* it is cleaned up when the pod dies (or in this case when the PV is deleted)
+
+We can create a new [PersistetVolumeSource](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L240) for EmptyDir. Users will request storage by creating PersistetVolumeClaims. If the PVC isn't bound to a PV, one will be created dynamically. The user can't control which node the storage for this PV comes from, or where in the node's filesystem it is allocated. Users can also hand create PVs bound to specific nodes (TODO: elaborate on QoS and what happens if there's no space on the node).
+
+## Scheduling constraints for data gravity
+
+The scheduler will schedule the pod paying attention to any referenced PVs. If an empty dir PV attached to an existing node is found, that trumps all scheduler predicates. Otherwise the scheduler makes a scheduling decision taking into account pod cpu/memory requests (and in the future, requested storage), attaches it to a node, and all subsequent scheduling decisions for any pod referencing the same PV evaluate to the same node. The scheduler "Attaches" a pv to a node by writing an annotation (TBD what actually writes this annotation, scheduler or kubelet, scheduler will at least bind the pod to the node). All pods referencing the PV that enter the "Running" state on the give node will be able to write to the volume. If the node enters the `NotReady` state, all pods referencing the PV attached to the node will remain in `Pending` till such time when an admin either removes the annotation, or rewrites it to point to another node. The directory allocated to the emptydir volume will be exported through the status field of the PV, so admins can yank the disk from a NotReady node and rsync data to the same directory on another node before rewriting the annotation.
+
+The annotations used by the scheduler to attach volumes to nodes will be: `volume.alpha.kubernetes.io/sticky-host`. The kubelet will reject any pod at admission time if it finds that the pod references a PV with this annotation set to a value other than its own node name.
+
+## Creation and Deletion
+
+The directory path allocated to an emptydir PV will be `/var/lib/kubelet/volumes/pv-uuid/`. This path will only get torn down when the PV is delete, or the `sticky-host` annotation on the PV is re-written to point to another node AND no pods are referencing the PV on the current node. It will NOT get deleted if all pods referencing the PV are deleted.
```

Comment:
s/delete/deleted/

> or the `sticky-host` annotation on the PV is re-written to point to another node

Hmm, what is actually supposed to happen in that case? Is this treated like creating a new emptyDir PV bound to a particular node? Why don't you just have validation forbid changing the sticky-host annotation on a PV? It seems like allowing it raises weird corner cases (like what happens to the pods that were referencing that PV? or do you only allow it when there are no pods referencing the PV?)